### ISSUE TITLE
Improved reading bufs reuse in port socket.

### ISSUE THREAD

**Issue by hongzhidao** [nginx/unit#206]
Hi.

It's common that reading data from port many times, thus after processing msg completely,
`msg.buf` is set to fmsg->buf, but it will be never equal to the current buf `b`, see line 593.
Finally these bufs will not be reused.

```
            /*
             * To disable instant completion or buffer re-usage,
             * handler should reset 'msg.buf'.
             */
            if (msg.buf == b) {
                nxt_port_buf_free(port, b);
            }
```

Fixed patch.

```
diff -r fd323ad9e24f src/nxt_port_socket.c
--- a/src/nxt_port_socket.c	Wed Dec 12 19:57:15 2018 +0300
+++ b/src/nxt_port_socket.c	Mon Dec 24 14:58:00 2018 -0500
@@ -797,7 +797,7 @@

                 port->handler(task, fmsg);

-                msg->buf = fmsg->buf;
+                nxt_port_buf_free(port, fmsg->buf);
                 msg->fd = fmsg->fd;
             }
         }
@@ -895,8 +895,7 @@
 static void
 nxt_port_buf_free(nxt_port_t *port, nxt_buf_t *b)
 {
-    b->next = port->free_bufs;
-    port->free_bufs = b;
+    nxt_buf_chain_add(&port->free_bufs, b);
 }
```

Thanks.

**Comment by mar0x** [comment]
Hi,

It is inaccurate to free (or recycle) buffer if handler explicitly request not to do it by resetting 'buf' field. I assume instant buffer completion should work even for fragmented messages because local 'b' variable is also altered prior to 'port->handler(..)' call.

Using 'nxt_buf_chain_add()' to recycle buffer is less efficient because it adds buffer to the end of the chain whereas order is not important in this case.

**Comment by hongzhidao** [comment]
@mar0x 

> It is inaccurate to free (or recycle) buffer if handler explicitly request not to do it by resetting 'buf' field. I assume instant buffer completion should work even for fragmented messages because local 'b' variable is also altered prior to 'port->handler(..)' call.

Sorry, I can't follow it.
1. For fragmented messages, should they be re-usage after completely processed?
```
794             if (nxt_fast_path(msg->port_msg.mf == 0)) {
795
796                 b = fmsg->buf;
797
798                 port->handler(task, fmsg);
799
800                 msg->buf = fmsg->buf;
801                 msg->fd = fmsg->fd;
802             }
```
  If yes, does port->handler need do something like 'msg->buf = ...'?
```
/*
             * To disable instant completion or buffer re-usage,
             * handler should reset 'msg.buf'.
             */
            if (msg.buf == b) {
                nxt_port_buf_free(port, b);
            }
```
Does the handler in 'handler should reset 'msg.buf'.' point to port->handler?

2.
> Using 'nxt_buf_chain_add()' to recycle buffer is less efficient because it adds buffer to the end of the chain whereas order is not important in this case.

```
static void
 nxt_port_buf_free(nxt_port_t *port, nxt_buf_t *b)
 {
-    b->next = port->free_bufs;
-    port->free_bufs = b;
+    nxt_buf_chain_add(&port->free_bufs, b);
 }
```
It's ok to append only one buf, but it does not work with buf has more next bufs.
```
@@ -797,7 +804,8 @@ nxt_port_read_msg_process(nxt_task_t *ta

                 port->handler(task, fmsg);

-                msg->buf = fmsg->buf;
+                //msg->buf = fmsg->buf;
+                nxt_port_buf_free(port, fmsg->buf);
                 msg->fd = fmsg->fd;
             }
```

3. Finally, I try to reproduce it, I'm not sure it's good with the hard-code.
Take a look, please.

version 1.
```
diff -r fd323ad9e24f src/nxt_port_socket.c
--- a/src/nxt_port_socket.c	Wed Dec 12 19:57:15 2018 +0300
+++ b/src/nxt_port_socket.c	Mon Feb 11 05:44:29 2019 +0800
@@ -80,8 +80,8 @@ nxt_port_socket_init(nxt_task_t *task, n
         }
     }

-    port->max_size = nxt_min(max_size, (size_t) sndbuf);
-    port->max_share = (64 * 1024);
+    port->max_size = 32;
+    port->max_share = 32;

     return NXT_OK;

@@ -594,6 +594,15 @@ nxt_port_read_handler(nxt_task_t *task,
                 nxt_port_buf_free(port, b);
             }

+printf("++++++++++\n");
+nxt_buf_t *buf;
+buf = port->free_bufs;
+while (buf != NULL) {
+    printf("%p\n", buf);
+    buf = buf->next;
+}
+printf("----------\n");
+
             if (port->socket.read_ready) {
                 continue;
             }
@@ -798,6 +807,7 @@ nxt_port_read_msg_process(nxt_task_t *ta
                 port->handler(task, fmsg);

                 msg->buf = fmsg->buf;
+                //nxt_port_buf_free(port, fmsg->buf);
                 msg->fd = fmsg->fd;
             }
         }
@@ -897,6 +907,7 @@ nxt_port_buf_free(nxt_port_t *port, nxt_
 {
     b->next = port->free_bufs;
     port->free_bufs = b;
+    //nxt_buf_chain_add(&port->free_bufs, b);
 }
```

version 2.
```
diff -r fd323ad9e24f src/nxt_port_socket.c
--- a/src/nxt_port_socket.c	Wed Dec 12 19:57:15 2018 +0300
+++ b/src/nxt_port_socket.c	Mon Feb 11 05:43:57 2019 +0800
@@ -80,8 +80,8 @@ nxt_port_socket_init(nxt_task_t *task, n
         }
     }

-    port->max_size = nxt_min(max_size, (size_t) sndbuf);
-    port->max_share = (64 * 1024);
+    port->max_size = 32;
+    port->max_share = 32;

     return NXT_OK;

@@ -594,6 +594,15 @@ nxt_port_read_handler(nxt_task_t *task,
                 nxt_port_buf_free(port, b);
             }

+printf("++++++++++\n");
+nxt_buf_t *buf;
+buf = port->free_bufs;
+while (buf != NULL) {
+    printf("%p\n", buf);
+    buf = buf->next;
+}
+printf("----------\n");
+
             if (port->socket.read_ready) {
                 continue;
             }
@@ -797,7 +806,8 @@ nxt_port_read_msg_process(nxt_task_t *ta

                 port->handler(task, fmsg);

-                msg->buf = fmsg->buf;
+                //msg->buf = fmsg->buf;
+                nxt_port_buf_free(port, fmsg->buf);
                 msg->fd = fmsg->fd;
             }
         }
@@ -895,8 +905,9 @@ nxt_port_buf_alloc(nxt_port_t *port)
 static void
 nxt_port_buf_free(nxt_port_t *port, nxt_buf_t *b)
 {
-    b->next = port->free_bufs;
-    port->free_bufs = b;
+    //b->next = port->free_bufs;
+    //port->free_bufs = b;
+    nxt_buf_chain_add(&port->free_bufs, b);
 }
```

Start unit without daemon, you can there are more re-usage bufs in version 2.

**Comment by mar0x** [comment]
> 1. For fragmented messages, should they be re-usage after completely processed?

Yes

> If yes, does port->handler need do something like 'msg->buf = ...'?

No

> If not, what's the reason?
> 
> ```
>             /*
>              * To disable instant completion or buffer re-usage,
>              * handler should reset 'msg.buf'.
>              */
>             if (msg.buf == b) {
>                 nxt_port_buf_free(port, b);
>             }
> ```

In the case handler wants to use buffer after processing and will free it by it's own, the handler should reset msg.buf. This is what written in comment above.

> Does the handler in 'handler should reset 'msg.buf'.' point to port->handler?

Yes. There is only one hander there.

> ```
> static void
>  nxt_port_buf_free(nxt_port_t *port, nxt_buf_t *b)
>  {
> -    b->next = port->free_bufs;
> -    port->free_bufs = b;
> +    nxt_buf_chain_add(&port->free_bufs, b);
>  }
> ```
> It's ok to append only one buf, but it does not work with buf has more next bufs.

This is true and needs to be fixed:

```
static void
nxt_port_buf_free(nxt_port_t *port, nxt_buf_t *b)
{
-    b->next = port->free_bufs;
+    nxt_buf_chain_add(&b, port->free_bufs);
     port->free_bufs = b;
}
```

I will try to apply your 'hard-code' patch to check.
Thank you!

**Comment by hongzhidao** [comment]
> ```
>  /*
>              * To disable instant completion or buffer re-usage,
>              * handler should reset 'msg.buf'.
>              */
>             if (msg.buf == b) {
>                 nxt_port_buf_free(port, b);
>             }
> ```
> In the case handler wants to use buffer after processing and will free it by it's own, the handler should reset msg.buf. This is what written in comment above.

@mar0x it's still kind of confusing, can you share more?
1. I think it's better that the handler (port->handler) needn't care msg.buf, and it doesn't know what to set.  `msg.buf = ?`
2. The condition avoid reuse the incomplete bufs, it' right. What I understand before is this, but it's different from the comment.
3. Can you make an example this case? 
> In the case handler wants to use buffer after processing

Thanks.

**Comment by mar0x** [comment]
> 1. I think it's better that the handler (port->handler) needn't care msg.buf, and it doesn't know what to set.  `msg.buf = ?`

If the handler "don't care" and does not change the msg.buf, the buffer will be recycled immediately after return from handler, because (msg.buf == b) condition will be TRUE. As far as I remember, all the current handlers does not change msg.buf. But if we have the handler which wants to postpone buffer re-use, such handler should reset msg.buf to NULL.

> 2. The condition avoid reusing incomplete bufs, it' right. What I understand before is this, but it's different from the comment.

I've wrote this condition only for further handler flexibility. Maybe it is redundant for now...

> 3. Can you make an example this case?

For example, handler which wants to send received buffer to other port of socket. In this case buffer should be freed only after send completion. So, handler store pointer to buffer locally, reset the msg.buf = NULL and do whatever it wants with this buffer, but after this it is responsible for proper buffer free.

> Thanks.

You are welcome! I'm highly appreciate your comments and questions too.

**Comment by hongzhidao** [comment]
Get it now, thanks.
1. It's a good idea to reserve the condition '(msg.buf == b)', it makes sense.
2. Except for the purpose, in the case of fragmented messages happens. It seems the buffers are not reused. As I showed.
```
--- a/src/nxt_port_socket.c	Wed Dec 12 19:57:15 2018 +0300
+++ b/src/nxt_port_socket.c	Mon Dec 24 14:58:00 2018 -0500
@@ -797,7 +797,7 @@

                 port->handler(task, fmsg);

                 msg->buf = fmsg->buf;
```
The condition never be satisfied.

**Comment by mar0x** [comment]
> 2. Except for the purpose, in the case of fragmented messages happens. It seems the buffers are not reused. As I showed, the condition never be satisfied.
> 
> ```
> --- a/src/nxt_port_socket.c	Wed Dec 12 19:57:15 2018 +0300
> +++ b/src/nxt_port_socket.c	Mon Dec 24 14:58:00 2018 -0500
> @@ -797,7 +797,7 @@
> 
>                  port->handler(task, fmsg);
> 
>                  msg->buf = fmsg->buf;
> ```

Again, this it because you miss one line in fragment above:

```
               ....
                b = fmsg->buf;

                port->handler(task, fmsg);

                msg->buf = fmsg->buf;
               ...

        if (msg->buf == b) {
        ...
```

msg->buf is changed after handler, but 'b' is also changed before. So, if handler does not change the fmsg->buf, condition (msg->buf == b) will be TRUE and msg->buf chain gets recycled (after nxt_port_buf_free() fix).

Does it makes sense?

**Comment by hongzhidao** [comment]
I download the latest source code. Take a look again, please.

In nxt_port_read_handler.
```
582         if (n > 0) {
583
584             msg.buf = b;
585             msg.size = n;
586
587             nxt_port_read_msg_process(task, port, &msg);
588
589             /*
590              * To disable instant completion or buffer re-usage,
591              * handler should reset 'msg.buf'.
592              */
593             if (msg.buf == b) {
594                 nxt_port_buf_free(port, b);
595             }
```
In nxt_port_read_msg_process.
```
794             if (nxt_fast_path(msg->port_msg.mf == 0)) {
795
796                 b = fmsg->buf;
797
798                 port->handler(task, fmsg);
799
800                 msg->buf = fmsg->buf;
801                 msg->fd = fmsg->fd;
802             }
```
In the case of fragmented messages.
fmsg->buf points to the first buf. The two local variable 'b' is different, so I think the condition will be false. Tell me if I misunderstand.

Well, It seems we are talking the differnt condition, there's another condition. So sorry.
In nxt_port_read_msg_process.
```
848 fmsg_failed:
849
850     if (msg->port_msg.mmap && orig_b != b) {
851
852         /*
853          * To disable instant buffer completion,
854          * handler should reset 'msg->buf'.
855          */
856         if (msg->buf == b) {
857             /* complete mmap buffers */
858             for (; b != NULL; b = b->next) {
859                 nxt_debug(task, "complete buffer %p", b);
860
861                 nxt_work_queue_add(port->socket.read_work_queue,
862                     b->completion_handler, task, b, b->parent);
863             }
864         }
865
866         /* restore original buf */
867         msg->buf = orig_b;
868     }
```
1. Will all the fragmented message buffers be freed in the label 'fmsg_failed'?
2. If yes, I need to deep into this part. If no, I think what I explained makes sense.

**Comment by mar0x** [comment]
You are right, fragmented (non-mmap) messages will not be reused. Yes, I've looked at the different part of the code. Here is proposed fix:

```
diff --git a/src/nxt_port_socket.c b/src/nxt_port_socket.c
--- a/src/nxt_port_socket.c
+++ b/src/nxt_port_socket.c
@@ -819,6 +819,14 @@ nxt_port_read_msg_process(nxt_task_t *ta
 
                 msg->buf = fmsg->buf;
                 msg->fd = fmsg->fd;
+
+                /*
+                 * To disable instant completion or buffer re-usage,
+                 * handler should reset 'msg.buf'.
+                 */
+                if (!msg->port_msg.mmap && msg.buf == b) {
+                    nxt_port_buf_free(port, b);
+                }
             }
         }
 
@@ -915,7 +923,7 @@ nxt_port_buf_alloc(nxt_port_t *port)
 static void
 nxt_port_buf_free(nxt_port_t *port, nxt_buf_t *b)
 {
-    b->next = port->free_bufs;
+    nxt_buf_chain_add(&b, port->free_bufs);
     port->free_bufs = b;
 }
```

First part will trigger reusing of fragmented chain if it is not in mmap and handler does not reset 'msg->buf'. Second part fixes reusing of buffer chains.

**Comment by hongzhidao** [comment]
@mar0x 
Great work.
```
static void
 nxt_port_buf_free(nxt_port_t *port, nxt_buf_t *b)
 {
-    b->next = port->free_bufs;
+    nxt_buf_chain_add(&b, port->free_bufs);
     port->free_bufs = b; /* This is unnecessary, how about deleting this line? */
 }
```

**Comment by mar0x** [comment]
```nxt_buf_chain_add(&b, port->free_bufs);```

will append current free buffers chain to the end of currently freed chain (b). port->free_bufs will refer at the middle of free buffers chain at this point and we need to adjust it to the first buffer in chain with assignment:

``` port->free_bufs = b;```

No?

**Comment by hongzhidao** [comment]
You are right, I didn't review it carefully.
But what about this?
```
nxt_buf_chain_add(&port->free_bufs, b);
```

**Comment by mar0x** [comment]
I assume ```b``` most likely is a single buffer chain, whereas ```port->free_bufs``` may have more than one spare buffers linked. This is the reason to append ```port->free_bufs``` to ```b```.

**Comment by hongzhidao** [comment]
Get it.
